hpy/devel/src/runtime/ctx_type.c

```diff
@@ -715,18 +934,41 @@ ctx_Type_FromSpec(HPyContext *ctx, HPyType_Spec *hpyspec,
     HPy_ssize_t base_member_offset;
     unsigned long flags = hpyspec->flags;
 
+    PyObject *bases = build_bases_from_params(params);
+    if (PyErr_Occurred()) {
+        PyMem_Free(spec->slots);
+        PyMem_Free(spec);
+        return HPy_NULL;
+    }
+
     bool is_pure;
     if (hpyspec->legacy != 0) {
         basicsize = hpyspec->basicsize;
         base_member_offset = 0;
         is_pure = false;
     }
     else {
-        // _HPy_PyObject_HEAD_SIZE ensures that the custom struct is
-        // correctly aligned.
         if (hpyspec->basicsize != 0) {
-            basicsize = hpyspec->basicsize + _HPy_PyObject_HEAD_SIZE;
-            base_member_offset = _HPy_PyObject_HEAD_SIZE;
+            if (bases) {
+                /* If base classes were specified, we need to compute the
+                   largest basicsize of all bases. We
+                   will then use the base's basicsize as offset for the pure
+                   HPy type's struct. In most cases, that will just be
+                   PyObject_HEAD but it could, e.g., also be PyHeapTypeObject
+                   in case of metaclasses. */
+                base_member_offset = compute_largest_basicsize(bases);
```

Comment:
I think there could be in theory an issue with memory alignment here that's why the other branch is using `_HPy_PyObject_HEAD_SIZE`, which is `sizeof` of some struct padded with extra field with a union of all possible types. We can probably change that union to a `max_align_t` field. In this branch we can maybe at least make the `base_member_offset` be divisible by `sizeof(max_align_t)`? Not sure if there is any more portable/less hacky solution.